File: sdk-py/primordia_sdk/mbs.py
"""
Machine Balance Sheet (MBS) v0.1
"""

from dataclasses import dataclass
from typing import List, Dict, Any, Optional, Tuple
import time

from .canonical import canonicalize_bytes
from .crypto import hash_bytes, sign, verify


MAX_SOLVENCY = 999999


@dataclass
class Asset:
    asset_type: str
    amount: int

    def to_dict(self) -> Dict[str, Any]:
        return {"asset_type": self.asset_type, "amount": self.amount}


@dataclass
class Liability:
    liability_type: str
    amount: int

    def to_dict(self) -> Dict[str, Any]:
        return {"liability_type": self.liability_type, "amount": self.amount}


@dataclass
class MBS:
    mbs_version: str
    agent_id: str
    assets: List[Asset]
    liabilities: List[Liability]
    burn_rate_usd_micros_per_s: int
    solvency_ratio: int
    timestamp_ms: int
    signature_ed25519: str

    def to_dict(self) -> Dict[str, Any]:
        return {
            "mbs_version": self.mbs_version,
            "agent_id": self.agent_id,
            "assets": [a.to_dict() for a in self.assets],
            "liabilities": [l.to_dict() for l in self.liabilities],
            "burn_rate_usd_micros_per_s": self.burn_rate_usd_micros_per_s,
            "solvency_ratio": self.solvency_ratio,
            "timestamp_ms": self.timestamp_ms,
            "signature_ed25519": self.signature_ed25519,
        }


def compute_solvency_ratio(assets: List[Asset], liabilities: List[Liability]) -> int:
    """Compute solvency ratio in basis points."""
    total_assets = sum(a.amount for a in assets)
    total_liabilities = sum(l.amount for l in liabilities)

    if total_liabilities == 0:
        return MAX_SOLVENCY

    return (total_assets * 10000) // total_liabilities


def compute_mbs(
    agent_id: str,
    assets: List[Asset],
    liabilities: List[Liability],
    burn_rate_usd_micros_per_s: int,
    private_key: str,
    timestamp_ms: Optional[int] = None,
) -> MBS:
    """Create a signed MBS."""
    solvency_ratio = compute_solvency_ratio(assets, liabilities)

    mbs_data = {
        "mbs_version": "0.1",
        "agent_id": agent_id,
        "assets": [a.to_dict() for a in assets],
        "liabilities": [l.to_dict() for l in liabilities],
        "burn_rate_usd_micros_per_s": burn_rate_usd_micros_per_s,
        "solvency_ratio": solvency_ratio,
        "timestamp_ms": timestamp_ms or int(time.time() * 1000),
    }

    canonical_bytes = canonicalize_bytes(mbs_data)
    mbs_hash = hash_bytes(canonical_bytes)
    signature = sign(mbs_hash, private_key)

    return MBS(
        mbs_version="0.1",
        agent_id=agent_id,
        assets=assets,
        liabilities=liabilities,
        burn_rate_usd_micros_per_s=burn_rate_usd_micros_per_s,
        solvency_ratio=solvency_ratio,
        timestamp_ms=mbs_data["timestamp_ms"],
        signature_ed25519=signature,
    )


def verify_mbs(mbs: MBS, public_key: str) -> Tuple[bool, Optional[str]]:
    """
    Verify an MBS.
    Returns (valid, error_message).
    """
    if mbs.mbs_version != "0.1":
        return False, "Invalid mbs_version"

    # Validate amounts
    for asset in mbs.assets:
        if asset.amount < 0:
            return False, "Asset amount cannot be negative"

    for liability in mbs.liabilities:
        if liability.amount < 0:
            return False, "Liability amount cannot be negative"

    if mbs.burn_rate_usd_micros_per_s < 0:
        return False, "Burn rate cannot be negative"

    # Verify solvency ratio
    expected_ratio = compute_solvency_ratio(mbs.assets, mbs.liabilities)
    if mbs.solvency_ratio != expected_ratio:
        return False, "Invalid solvency ratio"

    # Verify signature
    mbs_dict = mbs.to_dict()
    del mbs_dict["signature_ed25519"]
    mbs_hash = hash_bytes(canonicalize_bytes(mbs_dict))

    if not verify(mbs_hash, mbs.signature_ed25519, public_key):
        return False, "Invalid signature"

    return True, None


def compute_runway_seconds(mbs: MBS) -> float:
    """Compute runway in seconds until insolvency."""
    total_assets = sum(a.amount for a in mbs.assets)
    total_liabilities = sum(l.amount for l in mbs.liabilities)
    net_assets = total_assets - total_liabilities

    if mbs.burn_rate_usd_micros_per_s == 0:
        return float("inf") if net_assets > 0 else 0.0

    return max(0.0, net_assets / mbs.burn_rate_usd_micros_per_s)

File: sdk-ts/src/mbs.ts
/**
 * Machine Balance Sheet (MBS) v0.1
 */

import { canonicalizeBytes } from './canonical.js';
import { hash, sign, verify } from './crypto.js';

export interface Asset {
  asset_type: string;
  amount: number;
}

export interface Liability {
  liability_type: string;
  amount: number;
}

export interface MBS {
  mbs_version: '0.1';
  agent_id: string;
  assets: Asset[];
  liabilities: Liability[];
  burn_rate_usd_micros_per_s: number;
  solvency_ratio: number;
  timestamp_ms: number;
  signature_ed25519: string;
}

export interface MBSInput {
  agent_id: string;
  assets: Asset[];
  liabilities: Liability[];
  burn_rate_usd_micros_per_s: number;
  timestamp_ms?: number;
}

const MAX_SOLVENCY = 999999;

export function compute_solvency_ratio(assets: Asset[], liabilities: Liability[]): number {
  const totalAssets = assets.reduce((sum, a) => sum + a.amount, 0);
  const totalLiabilities = liabilities.reduce((sum, l) => sum + l.amount, 0);

  if (totalLiabilities === 0) {
    return MAX_SOLVENCY;
  }

  return Math.floor((totalAssets * 10000) / totalLiabilities);
}

export async function compute_mbs(input: MBSInput, privateKey: string): Promise<MBS> {
  const solvencyRatio = compute_solvency_ratio(input.assets, input.liabilities);

  const mbsWithoutSig = {
    mbs_version: '0.1' as const,
    agent_id: input.agent_id,
    assets: input.assets,
    liabilities: input.liabilities,
    burn_rate_usd_micros_per_s: input.burn_rate_usd_micros_per_s,
    solvency_ratio: solvencyRatio,
    timestamp_ms: input.timestamp_ms ?? Date.now()
  };

  const canonicalBytes = canonicalizeBytes(mbsWithoutSig);
  const mbsHash = hash(canonicalBytes);
  const signature = await sign(mbsHash, privateKey);

  return {
    ...mbsWithoutSig,
    signature_ed25519: signature
  };
}

export async function verify_mbs(mbs: MBS, publicKey: string): Promise<{ valid: boolean; error?: string }> {
  // Validate required fields
  if (mbs.mbs_version !== '0.1') {
    return { valid: false, error: 'Invalid mbs_version' };
  }

  // Validate asset amounts
  for (const asset of mbs.assets) {
    if (asset.amount < 0) {
      return { valid: false, error: 'Asset amount cannot be negative' };
    }
  }

  // Validate liability amounts
  for (const liability of mbs.liabilities) {
    if (liability.amount < 0) {
      return { valid: false, error: 'Liability amount cannot be negative' };
    }
  }

  // Validate burn rate
  if (mbs.burn_rate_usd_micros_per_s < 0) {
    return { valid: false, error: 'Burn rate cannot be negative' };
  }

  // Verify solvency ratio
  const expectedRatio = compute_solvency_ratio(mbs.assets, mbs.liabilities);
  if (mbs.solvency_ratio !== expectedRatio) {
    return { valid: false, error: 'Invalid solvency ratio' };
  }

  // Verify signature
  const { signature_ed25519, ...mbsWithoutSig } = mbs;
  const canonicalBytes = canonicalizeBytes(mbsWithoutSig);
  const mbsHash = hash(canonicalBytes);

  const isValid = await verify(mbsHash, signature_ed25519, publicKey);
  if (!isValid) {
    return { valid: false, error: 'Invalid signature' };
  }

  return { valid: true };
}

export function compute_runway_seconds(mbs: MBS): number {
  const totalAssets = mbs.assets.reduce((sum, a) => sum + a.amount, 0);
  const totalLiabilities = mbs.liabilities.reduce((sum, l) => sum + l.amount, 0);
  const netAssets = totalAssets - totalLiabilities;

  if (mbs.burn_rate_usd_micros_per_s === 0) {
    return netAssets > 0 ? Infinity : 0;
  }

  return Math.max(0, Math.floor(netAssets / mbs.burn_rate_usd_micros_per_s));
}
